test/bio_tfo_test.c

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/bio.h>
+#include "e_os.h"
+#include "internal/sockets.h"
+#include "testutil.h"
+
+/* If OS support is added in crypto/bio/bio_tfo.h, add it here */
+#if defined(OPENSSL_SYS_LINUX)
+# define GOOD_OS 1
+#elif defined(__FreeBSD__)
+# define GOOD_OS 1
+#elif defined(OPENSSL_SYS_MACOS)
+# define GOOD_OS 1
+#else
+# ifdef GOOD_OS
+#  undef GOOD_OS
+# endif
+#endif
+
+#if !defined(OPENSSL_NO_TFO) && defined(GOOD_OS)
+
+/*
+ * This test is to ensure that if TCP Fast Open is configured, that socket 
+ * connections will still work. These tests are able to detect if TCP Fast
+ * Open works, but the tests will pass as long as the socket connects.
+ *
+ * The first test function tests the socket interface as implemented as BIOs.
+ *
+ * The second test functions tests the socket interface as implemented as fds.
+ *
+ * The tests are run 5 times. The first time is without TFO.
+ * The second test will create the TCP fast open cookie,
+ * this can be seen in `ip tcp_metrics` and in /proc/net/netstat/ on Linux.
+ * e.g. on Linux 4.15.0-135-generic:
+ * $ grep '^TcpExt:' /proc/net/netstat | cut -d ' ' -f 84-90 | column -t
+ * The third attempt will use the cookie and actually do TCP fast open.
+ * The 4th time is client-TFO only, the 5th time is server-TFO only.
+ */
+
+#  define SOCKET_DATA "FooBar"
+#  define SOCKET_DATA_LEN sizeof(SOCKET_DATA)
+
+static int test_bio_tfo(int idx)
+{
+    BIO *cbio = NULL;
+    BIO *abio = NULL;
+    BIO *sbio = NULL;
+    int ret = 0;
+    int tmp;
+    int sockerr = 0;
+    const char *port;
+    int server_tfo = 0;
+    int client_tfo = 0;
+    size_t bytes;
+    char read_buffer[20];
+
+    switch (idx) {
+    default:
+    case 0:
+        break;
+    case 1:
+    case 2:
+        server_tfo = 1;
+        client_tfo = 1;
+        break;
+    case 3:
+        client_tfo = 1;
+        break;
+    case 4:
+        server_tfo = 1;
+        break;
+    }
+
+    /* ACCEPT SOCKET */
+    if (!TEST_ptr(abio = BIO_new_accept("127.0.0.1:0"))
```

Comment:
What will happen if we run this on an IPv6 only host?

Reply:
Not run? HA! Invariably, everything is dual-stack IPv4+IPv6 or just IPv4. I will update.

Reply:
The reason I ask is I *think* we ran into some oddball platforms like that with the TLSProxy code. I could be wrong though.

Reply:
It's now IPv4/IPv6 compatible.